File: packages/prism-core/src/layer1/plugins/platform/platform.ts
/**
 * @prism/plugin-platform — Platform Services Domain Registry (Layer 1)
 *
 * Registers calendar events, messages, reminders, and feeds.
 */

import type { EntityDef, EntityFieldDef, EdgeTypeDef } from "../../object-model/types.js";
import type { FluxAutomationPreset } from "../../flux/flux-types.js";
import { FLUX_TYPES } from "../../flux/flux-types.js";
import type { PrismPlugin } from "../../plugin/plugin-types.js";
import { pluginId } from "../../plugin/plugin-types.js";
import type { PlatformRegistry, PlatformEntityType, PlatformEdgeType } from "./platform-types.js";
import {
  PLATFORM_CATEGORIES, PLATFORM_TYPES, PLATFORM_EDGES,
  EVENT_RECURRENCES, MESSAGE_CHANNELS, REMINDER_PRIORITIES,
} from "./platform-types.js";

// ── Field Definitions ────────────────────────────────────────────────────

function enumOptions(values: ReadonlyArray<{ value: string; label: string }>): Array<{ value: string; label: string }> {
  return values.map(v => ({ value: v.value, label: v.label }));
}

const CALENDAR_EVENT_FIELDS: EntityFieldDef[] = [
  { id: "startTime", type: "datetime", label: "Start", required: true },
  { id: "endTime", type: "datetime", label: "End", required: true },
  { id: "allDay", type: "bool", label: "All Day", default: false },
  { id: "location", type: "string", label: "Location" },
  { id: "locationUrl", type: "url", label: "Location URL" },
  { id: "description", type: "text", label: "Description", ui: { multiline: true } },
  { id: "recurrence", type: "enum", label: "Recurrence", enumOptions: enumOptions(EVENT_RECURRENCES), default: "none" },
  { id: "recurrenceEnd", type: "date", label: "Recurrence End" },
  { id: "color", type: "color", label: "Color" },
  { id: "calendarId", type: "string", label: "Calendar ID" },
  { id: "externalId", type: "string", label: "External ID", ui: { hidden: true } },
  { id: "attendees", type: "string", label: "Attendees", ui: { placeholder: "comma-separated emails" } },
  { id: "remindMinutesBefore", type: "int", label: "Remind (min before)", default: 15 },
];

const MESSAGE_FIELDS: EntityFieldDef[] = [
  { id: "channel", type: "enum", label: "Channel", enumOptions: enumOptions(MESSAGE_CHANNELS), default: "internal" },
  { id: "from", type: "object_ref", label: "From", refTypes: [FLUX_TYPES.CONTACT] },
  { id: "to", type: "string", label: "To", ui: { placeholder: "recipients" } },
  { id: "subject", type: "string", label: "Subject" },
  { id: "body", type: "text", label: "Body", ui: { multiline: true } },
  { id: "sentAt", type: "datetime", label: "Sent At" },
  { id: "receivedAt", type: "datetime", label: "Received At" },
  { id: "threadId", type: "string", label: "Thread ID", ui: { hidden: true } },
  { id: "externalId", type: "string", label: "External ID", ui: { hidden: true } },
  { id: "isRead", type: "bool", label: "Read", default: false },
  { id: "isStarred", type: "bool", label: "Starred", default: false },
  { id: "hasAttachments", type: "bool", label: "Has Attachments", default: false },
];

const REMINDER_FIELDS: EntityFieldDef[] = [
  { id: "dueAt", type: "datetime", label: "Due At", required: true },
  { id: "priority", type: "enum", label: "Priority", enumOptions: enumOptions(REMINDER_PRIORITIES), default: "normal" },
  { id: "snoozedUntil", type: "datetime", label: "Snoozed Until" },
  { id: "recurring", type: "enum", label: "Recurring", enumOptions: enumOptions(EVENT_RECURRENCES), default: "none" },
  { id: "description", type: "text", label: "Description", ui: { multiline: true } },
];

const FEED_FIELDS: EntityFieldDef[] = [
  { id: "feedUrl", type: "url", label: "Feed URL", required: true },
  { id: "feedType", type: "enum", label: "Feed Type", enumOptions: [
    { value: "rss", label: "RSS" },
    { value: "atom", label: "Atom" },
    { value: "json", label: "JSON Feed" },
    { value: "webhook", label: "Webhook" },
  ] },
  { id: "refreshIntervalMinutes", type: "int", label: "Refresh Interval (min)", default: 60 },
  { id: "lastFetchedAt", type: "datetime", label: "Last Fetched", ui: { readonly: true } },
  { id: "itemCount", type: "int", label: "Items", default: 0, ui: { readonly: true } },
  { id: "autoArchiveDays", type: "int", label: "Auto-archive After (days)", default: 30 },
];

const FEED_ITEM_FIELDS: EntityFieldDef[] = [
  { id: "title", type: "string", label: "Title" },
  { id: "url", type: "url", label: "URL" },
  { id: "author", type: "string", label: "Author" },
  { id: "publishedAt", type: "datetime", label: "Published At" },
  { id: "summary", type: "text", label: "Summary", ui: { multiline: true } },
  { id: "content", type: "text", label: "Content", ui: { multiline: true } },
  { id: "isRead", type: "bool", label: "Read", default: false },
  { id: "isStarred", type: "bool", label: "Starred", default: false },
  { id: "externalId", type: "string", label: "External ID", ui: { hidden: true } },
];

// ── Entity Definitions ───────────────────────────────────────────────────

function buildEntityDefs(): EntityDef[] {
  return [
    {
      type: PLATFORM_TYPES.CALENDAR_EVENT,
      nsid: "io.prismapp.platform.calendar-event",
      category: PLATFORM_CATEGORIES.CALENDAR,
      label: "Event",
      pluralLabel: "Events",
      defaultChildView: "timeline",
      fields: CALENDAR_EVENT_FIELDS,
    },
    {
      type: PLATFORM_TYPES.MESSAGE,
      nsid: "io.prismapp.platform.message",
      category: PLATFORM_CATEGORIES.MESSAGING,
      label: "Message",
      pluralLabel: "Messages",
      defaultChildView: "list",
      fields: MESSAGE_FIELDS,
    },
    {
      type: PLATFORM_TYPES.REMINDER,
      nsid: "io.prismapp.platform.reminder",
      category: PLATFORM_CATEGORIES.REMINDERS,
      label: "Reminder",
      pluralLabel: "Reminders",
      defaultChildView: "list",
      fields: REMINDER_FIELDS,
    },
    {
      type: PLATFORM_TYPES.FEED,
      nsid: "io.prismapp.platform.feed",
      category: PLATFORM_CATEGORIES.FEEDS,
      label: "Feed",
      pluralLabel: "Feeds",
      defaultChildView: "list",
      fields: FEED_FIELDS,
      extraChildTypes: [PLATFORM_TYPES.FEED_ITEM],
    },
    {
      type: PLATFORM_TYPES.FEED_ITEM,
      nsid: "io.prismapp.platform.feed-item",
      category: PLATFORM_CATEGORIES.FEEDS,
      label: "Feed Item",
      pluralLabel: "Feed Items",
      childOnly: true,
      fields: FEED_ITEM_FIELDS,
    },
  ];
}

// ── Edge Definitions ─────────────────────────────────────────────────────

function buildEdgeDefs(): EdgeTypeDef[] {
  return [
    {
      relation: PLATFORM_EDGES.REMINDS_ABOUT,
      nsid: "io.prismapp.platform.reminds-about",
      label: "Reminds About",
      behavior: "weak",
      sourceTypes: [PLATFORM_TYPES.REMINDER],
      description: "Links a reminder to any object",
      suggestInline: true,
    },
    {
      relation: PLATFORM_EDGES.EVENT_FOR,
      nsid: "io.prismapp.platform.event-for",
      label: "Event For",
      behavior: "weak",
      sourceTypes: [PLATFORM_TYPES.CALENDAR_EVENT],
      targetTypes: [FLUX_TYPES.PROJECT, FLUX_TYPES.CONTACT, FLUX_TYPES.ORGANIZATION],
      suggestInline: true,
    },
    {
      relation: PLATFORM_EDGES.REPLY_TO,
      nsid: "io.prismapp.platform.reply-to",
      label: "Reply To",
      behavior: "weak",
      sourceTypes: [PLATFORM_TYPES.MESSAGE],
      targetTypes: [PLATFORM_TYPES.MESSAGE],
    },
    {
      relation: PLATFORM_EDGES.FEED_SOURCE,
      nsid: "io.prismapp.platform.feed-source",
      label: "Feed Source",
      behavior: "membership",
      sourceTypes: [PLATFORM_TYPES.FEED_ITEM],
      targetTypes: [PLATFORM_TYPES.FEED],
    },
  ];
}

// ── Automation Presets ────────────────────────────────────────────────────

function buildAutomationPresets(): FluxAutomationPreset[] {
  return [
    {
      id: "platform:auto:reminder-due",
      name: "Fire reminder notification",
      entityType: PLATFORM_TYPES.REMINDER,
      trigger: "on_due_date",
      condition: "status == 'pending'",
      actions: [
        { kind: "send_notification", target: "owner", value: "Reminder: {{name}}" },
      ],
    },
    {
      id: "platform:auto:event-reminder",
      name: "Pre-event reminder",
      entityType: PLATFORM_TYPES.CALENDAR_EVENT,
      trigger: "on_schedule",
      condition: "remindMinutesBefore > 0",
      actions: [
        { kind: "send_notification", target: "owner", value: "Event '{{name}}' starting in {{remindMinutesBefore}} minutes" },
      ],
    },
    {
      id: "platform:auto:feed-archive",
      name: "Auto-archive old feed items",
      entityType: PLATFORM_TYPES.FEED_ITEM,
      trigger: "on_schedule",
      condition: "daysOld > parent.autoArchiveDays",
      actions: [
        { kind: "move_to_status", target: "status", value: "archived" },
      ],
    },
  ];
}

// ── Plugin ───────────────────────────────────────────────────────────────

function buildPlugin(): PrismPlugin {
  return {
    id: pluginId("prism.plugin.platform"),
    name: "Platform",
    contributes: {
      views: [
        { id: "platform:calendar", label: "Calendar", zone: "content", componentId: "CalendarView", description: "Calendar and events" },
        { id: "platform:inbox", label: "Inbox", zone: "content", componentId: "InboxView", description: "Unified messaging inbox" },
        { id: "platform:reminders", label: "Reminders", zone: "content", componentId: "ReminderListView", description: "Reminder manager" },
        { id: "platform:feeds", label: "Feeds", zone: "content", componentId: "FeedReaderView", description: "News & feed reader" },
      ],
      commands: [
        { id: "platform:new-event", label: "New Event", category: "Platform", action: "platform.newEvent" },
        { id: "platform:new-reminder", label: "New Reminder", category: "Platform", action: "platform.newReminder" },
        { id: "platform:compose-message", label: "Compose Message", category: "Platform", action: "platform.composeMessage" },
        { id: "platform:add-feed", label: "Add Feed", category: "Platform", action: "platform.addFeed" },
      ],
      keybindings: [
        { command: "platform:new-event", key: "ctrl+shift+e" },
        { command: "platform:new-reminder", key: "ctrl+shift+r" },
      ],
      activityBar: [
        { id: "platform:activity", label: "Platform", position: "top", priority: 10 },
      ],
    },
  };
}

// ── Factory ──────────────────────────────────────────────────────────────

export function createPlatformRegistry(): PlatformRegistry {
  const entityDefs = buildEntityDefs();
  const edgeDefs = buildEdgeDefs();
  const presets = buildAutomationPresets();
  const plugin = buildPlugin();

  return {
    getEntityDefs: () => entityDefs,
    getEdgeDefs: () => edgeDefs,
    getEntityDef: (type: PlatformEntityType) => entityDefs.find(d => d.type === type),
    getEdgeDef: (relation: PlatformEdgeType) => edgeDefs.find(d => d.relation === relation),
    getAutomationPresets: () => presets,
    getPlugin: () => plugin,
  };
}

// ── Self-Registering Bundle ──────────────────────────────────────────────

import type { PluginBundle, PluginInstallContext } from "../plugin-install.js";

export function createPlatformBundle(): PluginBundle {
  return {
    id: "prism.plugin.platform",
    name: "Platform",
    install(ctx: PluginInstallContext) {
      const reg = createPlatformRegistry();
      ctx.objectRegistry.registerAll(reg.getEntityDefs());
      ctx.objectRegistry.registerEdges(reg.getEdgeDefs());
      return ctx.pluginRegistry.register(reg.getPlugin());
    },
  };
}

File: packages/prism-core/src/kernel/plugin_bundles/platform.rs
//! `plugin_bundles::platform` — Platform services bundle.
//!
//! Port of `kernel/plugin-bundles/platform/platform.ts`: calendar
//! events, messages, reminders, and feeds. Cross-cutting integrations
//! rather than a single domain.

use serde_json::json;

use super::builders::{
    edge_def, entity_def, enum_options, owned_strings, ui_hidden, ui_multiline, ui_placeholder,
    ui_readonly, EdgeSpec, EntitySpec, Field,
};
use super::flux_types::{
    flux_types, FluxActionKind, FluxAutomationAction, FluxAutomationPreset, FluxTriggerKind,
};
use super::install::{PluginBundle, PluginInstallContext};
use crate::foundation::object_model::types::DefaultChildView;
use crate::foundation::object_model::{
    EdgeBehavior, EdgeTypeDef, EntityDef, EntityFieldDef, EntityFieldType, EnumOption,
};
use crate::kernel::plugin::{
    plugin_id, ActivityBarContributionDef, ActivityBarPosition, CommandContributionDef,
    KeybindingContributionDef, PluginContributions, PrismPlugin, ViewContributionDef, ViewZone,
};

// ── Domain constants ────────────────────────────────────────────────────────

pub mod platform_categories {
    pub const CALENDAR: &str = "platform:calendar";
    pub const MESSAGING: &str = "platform:messaging";
    pub const REMINDERS: &str = "platform:reminders";
    pub const FEEDS: &str = "platform:feeds";
}

pub mod platform_types {
    pub const CALENDAR_EVENT: &str = "platform:calendar-event";
    pub const MESSAGE: &str = "platform:message";
    pub const REMINDER: &str = "platform:reminder";
    pub const FEED: &str = "platform:feed";
    pub const FEED_ITEM: &str = "platform:feed-item";
}

pub mod platform_edges {
    pub const REMINDS_ABOUT: &str = "platform:reminds-about";
    pub const EVENT_FOR: &str = "platform:event-for";
    pub const REPLY_TO: &str = "platform:reply-to";
    pub const FEED_SOURCE: &str = "platform:feed-source";
}

// ── Option tables ───────────────────────────────────────────────────────────

fn event_recurrences() -> Vec<EnumOption> {
    enum_options(&[
        ("none", "None"),
        ("daily", "Daily"),
        ("weekly", "Weekly"),
        ("biweekly", "Bi-weekly"),
        ("monthly", "Monthly"),
        ("yearly", "Yearly"),
    ])
}

fn message_channels() -> Vec<EnumOption> {
    enum_options(&[
        ("internal", "Internal"),
        ("email", "Email"),
        ("sms", "SMS"),
        ("slack", "Slack"),
        ("discord", "Discord"),
    ])
}

fn reminder_priorities() -> Vec<EnumOption> {
    enum_options(&[
        ("low", "Low"),
        ("normal", "Normal"),
        ("high", "High"),
        ("urgent", "Urgent"),
    ])
}

// ── Fields ──────────────────────────────────────────────────────────────────

fn calendar_event_fields() -> Vec<EntityFieldDef> {
    vec![
        Field::new("startTime", EntityFieldType::Datetime)
            .label("Start")
            .required()
            .build(),
        Field::new("endTime", EntityFieldType::Datetime)
            .label("End")
            .required()
            .build(),
        Field::new("allDay", EntityFieldType::Bool)
            .label("All Day")
            .default(json!(false))
            .build(),
        Field::new("location", EntityFieldType::String)
            .label("Location")
            .build(),
        Field::new("locationUrl", EntityFieldType::Url)
            .label("Location URL")
            .build(),
        Field::new("description", EntityFieldType::Text)
            .label("Description")
            .ui(ui_multiline())
            .build(),
        Field::new("recurrence", EntityFieldType::Enum)
            .label("Recurrence")
            .enum_values(event_recurrences())
            .default(json!("none"))
            .build(),
        Field::new("recurrenceEnd", EntityFieldType::Date)
            .label("Recurrence End")
            .build(),
        Field::new("color", EntityFieldType::Color)
            .label("Color")
            .build(),
        Field::new("calendarId", EntityFieldType::String)
            .label("Calendar ID")
            .build(),
        Field::new("externalId", EntityFieldType::String)
            .label("External ID")
            .ui(ui_hidden())
            .build(),
        Field::new("attendees", EntityFieldType::String)
            .label("Attendees")
            .ui(ui_placeholder("comma-separated emails"))
            .build(),
        Field::new("remindMinutesBefore", EntityFieldType::Int)
            .label("Remind (min before)")
            .default(json!(15))
            .build(),
    ]
}

fn message_fields() -> Vec<EntityFieldDef> {
    vec![
        Field::new("channel", EntityFieldType::Enum)
            .label("Channel")
            .enum_values(message_channels())
            .default(json!("internal"))
            .build(),
        Field::new("from", EntityFieldType::ObjectRef)
            .label("From")
            .ref_types([flux_types::CONTACT])
            .build(),
        Field::new("to", EntityFieldType::String)
            .label("To")
            .ui(ui_placeholder("recipients"))
            .build(),
        Field::new("subject", EntityFieldType::String)
            .label("Subject")
            .build(),
        Field::new("body", EntityFieldType::Text)
            .label("Body")
            .ui(ui_multiline())
            .build(),
        Field::new("sentAt", EntityFieldType::Datetime)
            .label("Sent At")
            .build(),
        Field::new("receivedAt", EntityFieldType::Datetime)
            .label("Received At")
            .build(),
        Field::new("threadId", EntityFieldType::String)
            .label("Thread ID")
            .ui(ui_hidden())
            .build(),
        Field::new("externalId", EntityFieldType::String)
            .label("External ID")
            .ui(ui_hidden())
            .build(),
        Field::new("isRead", EntityFieldType::Bool)
            .label("Read")
            .default(json!(false))
            .build(),
        Field::new("isStarred", EntityFieldType::Bool)
            .label("Starred")
            .default(json!(false))
            .build(),
        Field::new("hasAttachments", EntityFieldType::Bool)
            .label("Has Attachments")
            .default(json!(false))
            .build(),
    ]
}

fn reminder_fields() -> Vec<EntityFieldDef> {
    vec![
        Field::new("dueAt", EntityFieldType::Datetime)
            .label("Due At")
            .required()
            .build(),
        Field::new("priority", EntityFieldType::Enum)
            .label("Priority")
            .enum_values(reminder_priorities())
            .default(json!("normal"))
            .build(),
        Field::new("snoozedUntil", EntityFieldType::Datetime)
            .label("Snoozed Until")
            .build(),
        Field::new("recurring", EntityFieldType::Enum)
            .label("Recurring")
            .enum_values(event_recurrences())
            .default(json!("none"))
            .build(),
        Field::new("description", EntityFieldType::Text)
            .label("Description")
            .ui(ui_multiline())
            .build(),
    ]
}

fn feed_fields() -> Vec<EntityFieldDef> {
    vec![
        Field::new("feedUrl", EntityFieldType::Url)
            .label("Feed URL")
            .required()
            .build(),
        Field::new("feedType", EntityFieldType::Enum)
            .label("Feed Type")
            .enum_values(enum_options(&[
                ("rss", "RSS"),
                ("atom", "Atom"),
                ("json", "JSON Feed"),
                ("webhook", "Webhook"),
            ]))
            .build(),
        Field::new("refreshIntervalMinutes", EntityFieldType::Int)
            .label("Refresh Interval (min)")
            .default(json!(60))
            .build(),
        Field::new("lastFetchedAt", EntityFieldType::Datetime)
            .label("Last Fetched")
            .ui(ui_readonly())
            .build(),
        Field::new("itemCount", EntityFieldType::Int)
            .label("Items")
            .default(json!(0))
            .ui(ui_readonly())
            .build(),
        Field::new("autoArchiveDays", EntityFieldType::Int)
            .label("Auto-archive After (days)")
            .default(json!(30))
            .build(),
    ]
}

fn feed_item_fields() -> Vec<EntityFieldDef> {
    vec![
        Field::new("title", EntityFieldType::String)
            .label("Title")
            .build(),
        Field::new("url", EntityFieldType::Url).label("URL").build(),
        Field::new("author", EntityFieldType::String)
            .label("Author")
            .build(),
        Field::new("publishedAt", EntityFieldType::Datetime)
            .label("Published At")
            .build(),
        Field::new("summary", EntityFieldType::Text)
            .label("Summary")
            .ui(ui_multiline())
            .build(),
        Field::new("content", EntityFieldType::Text)
            .label("Content")
            .ui(ui_multiline())
            .build(),
        Field::new("isRead", EntityFieldType::Bool)
            .label("Read")
            .default(json!(false))
            .build(),
        Field::new("isStarred", EntityFieldType::Bool)
            .label("Starred")
            .default(json!(false))
            .build(),
        Field::new("externalId", EntityFieldType::String)
            .label("External ID")
            .ui(ui_hidden())
            .build(),
    ]
}

// ── Entity + edge defs ──────────────────────────────────────────────────────

pub fn build_entity_defs() -> Vec<EntityDef> {
    vec![
        entity_def(EntitySpec {
            type_name: platform_types::CALENDAR_EVENT,
            nsid: "io.prismapp.platform.calendar-event",
            category: platform_categories::CALENDAR,
            label: "Event",
            plural_label: "Events",
            default_child_view: Some(DefaultChildView::Timeline),
            child_only: false,
            extra_child_types: None,
            fields: calendar_event_fields(),
        }),
        entity_def(EntitySpec {
            type_name: platform_types::MESSAGE,
            nsid: "io.prismapp.platform.message",
            category: platform_categories::MESSAGING,
            label: "Message",
            plural_label: "Messages",
            default_child_view: Some(DefaultChildView::List),
            child_only: false,
            extra_child_types: None,
            fields: message_fields(),
        }),
        entity_def(EntitySpec {
            type_name: platform_types::REMINDER,
            nsid: "io.prismapp.platform.reminder",
            category: platform_categories::REMINDERS,
            label: "Reminder",
            plural_label: "Reminders",
            default_child_view: Some(DefaultChildView::List),
            child_only: false,
            extra_child_types: None,
            fields: reminder_fields(),
        }),
        entity_def(EntitySpec {
            type_name: platform_types::FEED,
            nsid: "io.prismapp.platform.feed",
            category: platform_categories::FEEDS,
            label: "Feed",
            plural_label: "Feeds",
            default_child_view: Some(DefaultChildView::List),
            child_only: false,
            extra_child_types: Some(owned_strings([platform_types::FEED_ITEM])),
            fields: feed_fields(),
        }),
        entity_def(EntitySpec {
            type_name: platform_types::FEED_ITEM,
            nsid: "io.prismapp.platform.feed-item",
            category: platform_categories::FEEDS,
            label: "Feed Item",
            plural_label: "Feed Items",
            default_child_view: None,
            child_only: true,
            extra_child_types: None,
            fields: feed_item_fields(),
        }),
    ]
}

pub fn build_edge_defs() -> Vec<EdgeTypeDef> {
    vec![
        edge_def(EdgeSpec {
            relation: platform_edges::REMINDS_ABOUT,
            nsid: "io.prismapp.platform.reminds-about",
            label: "Reminds About",
            behavior: EdgeBehavior::Weak,
            source_types: owned_strings([platform_types::REMINDER]),
            target_types: None,
            description: Some("Links a reminder to any object"),
            suggest_inline: true,
            undirected: false,
        }),
        edge_def(EdgeSpec {
            relation: platform_edges::EVENT_FOR,
            nsid: "io.prismapp.platform.event-for",
            label: "Event For",
            behavior: EdgeBehavior::Weak,
            source_types: owned_strings([platform_types::CALENDAR_EVENT]),
            target_types: Some(owned_strings([
                flux_types::PROJECT,
                flux_types::CONTACT,
                flux_types::ORGANIZATION,
            ])),
            description: None,
            suggest_inline: true,
            undirected: false,
        }),
        edge_def(EdgeSpec {
            relation: platform_edges::REPLY_TO,
            nsid: "io.prismapp.platform.reply-to",
            label: "Reply To",
            behavior: EdgeBehavior::Weak,
            source_types: owned_strings([platform_types::MESSAGE]),
            target_types: Some(owned_strings([platform_types::MESSAGE])),
            description: None,
            suggest_inline: false,
            undirected: false,
        }),
        edge_def(EdgeSpec {
            relation: platform_edges::FEED_SOURCE,
            nsid: "io.prismapp.platform.feed-source",
            label: "Feed Source",
            behavior: EdgeBehavior::Membership,
            source_types: owned_strings([platform_types::FEED_ITEM]),
            target_types: Some(owned_strings([platform_types::FEED])),
            description: None,
            suggest_inline: false,
            undirected: false,
        }),
    ]
}

// ── Automation presets ──────────────────────────────────────────────────────

pub fn build_automation_presets() -> Vec<FluxAutomationPreset> {
    vec![
        FluxAutomationPreset {
            id: "platform:auto:reminder-due".into(),
            name: "Fire reminder notification".into(),
            entity_type: platform_types::REMINDER.into(),
            trigger: FluxTriggerKind::OnDueDate,
            condition: Some("status == 'pending'".into()),
            actions: vec![FluxAutomationAction {
                kind: FluxActionKind::SendNotification,
                target: "owner".into(),
                value: "Reminder: {{name}}".into(),
            }],
        },
        FluxAutomationPreset {
            id: "platform:auto:event-reminder".into(),
            name: "Pre-event reminder".into(),
            entity_type: platform_types::CALENDAR_EVENT.into(),
            trigger: FluxTriggerKind::OnSchedule,
            condition: Some("remindMinutesBefore > 0".into()),
            actions: vec![FluxAutomationAction {
                kind: FluxActionKind::SendNotification,
                target: "owner".into(),
                value: "Event '{{name}}' starting in {{remindMinutesBefore}} minutes".into(),
            }],
        },
        FluxAutomationPreset {
            id: "platform:auto:feed-archive".into(),
            name: "Auto-archive old feed items".into(),
            entity_type: platform_types::FEED_ITEM.into(),
            trigger: FluxTriggerKind::OnSchedule,
            condition: Some("daysOld > parent.autoArchiveDays".into()),
            actions: vec![FluxAutomationAction {
                kind: FluxActionKind::MoveToStatus,
                target: "status".into(),
                value: "archived".into(),
            }],
        },
    ]
}

// ── Plugin ──────────────────────────────────────────────────────────────────

fn view(id: &str, label: &str, component_id: &str, description: &str) -> ViewContributionDef {
    ViewContributionDef {
        id: id.into(),
        label: label.into(),
        zone: ViewZone::Content,
        component_id: component_id.into(),
        icon: None,
        default_visible: None,
        description: Some(description.into()),
        tags: None,
    }
}

fn command(id: &str, label: &str, action: &str) -> CommandContributionDef {
    CommandContributionDef {
        id: id.into(),
        label: label.into(),
        category: "Platform".into(),
        shortcut: None,
        description: None,
        action: action.into(),
        payload: None,
        when: None,
    }
}

pub fn build_plugin() -> PrismPlugin {
    PrismPlugin::new(plugin_id("prism.plugin.platform"), "Platform").with_contributes(
        PluginContributions {
            views: Some(vec![
                view(
                    "platform:calendar",
                    "Calendar",
                    "CalendarView",
                    "Calendar and events",
                ),
                view(
                    "platform:inbox",
                    "Inbox",
                    "InboxView",
                    "Unified messaging inbox",
                ),
                view(
                    "platform:reminders",
                    "Reminders",
                    "ReminderListView",
                    "Reminder manager",
                ),
                view(
                    "platform:feeds",
                    "Feeds",
                    "FeedReaderView",
                    "News & feed reader",
                ),
            ]),
            commands: Some(vec![
                command("platform:new-event", "New Event", "platform.newEvent"),
                command(
                    "platform:new-reminder",
                    "New Reminder",
                    "platform.newReminder",
                ),
                command(
                    "platform:compose-message",
                    "Compose Message",
                    "platform.composeMessage",
                ),
                command("platform:add-feed", "Add Feed", "platform.addFeed"),
            ]),
            keybindings: Some(vec![
                KeybindingContributionDef {
                    command: "platform:new-event".into(),
                    key: "ctrl+shift+e".into(),
                    when: None,
                },
                KeybindingContributionDef {
                    command: "platform:new-reminder".into(),
                    key: "ctrl+shift+r".into(),
                    when: None,
                },
            ]),
            activity_bar: Some(vec![ActivityBarContributionDef {
                id: "platform:activity".into(),
                label: "Platform".into(),
                icon: None,
                position: Some(ActivityBarPosition::Top),
                priority: Some(10),
            }]),
            context_menus: None,
            settings: None,
            toolbar: None,
            status_bar: None,
            weak_ref_providers: None,
            immersive: None,
        },
    )
}

pub struct PlatformBundle;

impl PluginBundle for PlatformBundle {
    fn id(&self) -> &str {
        "prism.plugin.platform"
    }

    fn name(&self) -> &str {
        "Platform"
    }

    fn install(&self, ctx: &mut PluginInstallContext<'_>) {
        ctx.object_registry.register_all(build_entity_defs());
        ctx.object_registry.register_edges(build_edge_defs());
        ctx.plugin_registry.register(build_plugin());
    }
}

pub fn create_platform_bundle() -> Box<dyn PluginBundle> {
    Box::new(PlatformBundle)
}
